pkg/resources/helpers.go

```diff
@@ -330,3 +330,183 @@ func ListDiff[T comparable](beforeList []T, afterList []T) (added []T, removed [
+			s1.S3StorageLocationParams.StorageProvider == s2.S3StorageLocationParams.StorageProvider &&
+			s1.S3StorageLocationParams.StorageAwsRoleArn == s2.S3StorageLocationParams.StorageAwsRoleArn &&
+			s1.S3StorageLocationParams.StorageBaseUrl == s2.S3StorageLocationParams.StorageBaseUrl {
+
+			if s1.S3StorageLocationParams.Encryption == nil && s2.S3StorageLocationParams.Encryption == nil {
+				return true, nil
+			} else if s1.S3StorageLocationParams.Encryption != nil && s2.S3StorageLocationParams.Encryption != nil {
+				typeEqual := s1.S3StorageLocationParams.Encryption.Type == s2.S3StorageLocationParams.Encryption.Type
+				kmsKeyIdEqual := s1.S3StorageLocationParams.Encryption.KmsKeyId == nil && s2.S3StorageLocationParams.Encryption.KmsKeyId == nil ||
+					(s1.S3StorageLocationParams.Encryption.KmsKeyId != nil && s2.S3StorageLocationParams.Encryption.KmsKeyId != nil &&
+						*s1.S3StorageLocationParams.Encryption.KmsKeyId == *s2.S3StorageLocationParams.Encryption.KmsKeyId)
+				return typeEqual && kmsKeyIdEqual, nil
+			}
+		}
+	case sdk.StorageProviderGCS:
+		if s1.GCSStorageLocationParams.Name == s2.GCSStorageLocationParams.Name &&
+			s1.GCSStorageLocationParams.StorageBaseUrl == s2.GCSStorageLocationParams.StorageBaseUrl {
+
+			if s1.GCSStorageLocationParams.Encryption == nil && s2.GCSStorageLocationParams.Encryption == nil {
+				return true, nil
+			} else if s1.GCSStorageLocationParams.Encryption != nil && s2.GCSStorageLocationParams.Encryption != nil {
+				typeEqual := s1.GCSStorageLocationParams.Encryption.Type == s2.GCSStorageLocationParams.Encryption.Type
+				kmsKeyIdEqual := s1.GCSStorageLocationParams.Encryption.KmsKeyId == nil && s2.GCSStorageLocationParams.Encryption.KmsKeyId == nil ||
+					(s1.GCSStorageLocationParams.Encryption.KmsKeyId != nil && s2.GCSStorageLocationParams.Encryption.KmsKeyId != nil &&
+						*s1.GCSStorageLocationParams.Encryption.KmsKeyId == *s2.GCSStorageLocationParams.Encryption.KmsKeyId)
+				return typeEqual && kmsKeyIdEqual, nil
+			}
+		}
+	case sdk.StorageProviderAzure:
+		return s1.AzureStorageLocationParams.Name == s2.AzureStorageLocationParams.Name &&
+			s1.AzureStorageLocationParams.AzureTenantId == s2.AzureStorageLocationParams.AzureTenantId &&
+			s1.AzureStorageLocationParams.StorageBaseUrl == s2.AzureStorageLocationParams.StorageBaseUrl, nil
+	}
+
+	return false, nil
+}
+
+// Returns a copy of the given storage location with a 'temp_' prefix on the Name field
+func CopyStorageLocationWithTempName(
+	storageLocation sdk.ExternalVolumeStorageLocation,
+) (sdk.ExternalVolumeStorageLocation, error) {
+	storageProvider, err := GetStorageLocationStorageProvider(storageLocation)
+	if err != nil {
+		return sdk.ExternalVolumeStorageLocation{}, err
+	}
+
+	currName, err := GetStorageLocationName(storageLocation)
+	if err != nil {
+		return sdk.ExternalVolumeStorageLocation{}, err
+	}
+
+	newName := fmt.Sprintf("temp_%s", currName)
```

Comment:
It's rather unlikely, but what if someone actually has this name in Snowflake? If we decide to stick with the copying implementation, we should write about this behavior in the docs.

Reply:
Shouldn't be possible as storage location names only need to be unique per external volume and we're removing all storage locations except the first, and then adding a prefix to the name of this first - https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/afbd4a00ec193042456b0f2a2c1bf99b3b494f4a/pkg/resources/external_volume.go#L659.

Reply:
Note that only happens when necessary too - if you've only changed the last item in the storage location list for example, a temporary won't need to be created - https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/afbd4a00ec193042456b0f2a2c1bf99b3b494f4a/pkg/resources/external_volume.go#L694

Reply:
Cool. So I'd change `temp_` to something more specific, like `terraform_provider_copy_`, and write about this in the docs (the whole behavior and the prefix we use). I think you can add this in `templates/resources/external_volume.md.tmpl`.

Reply:
Sounds good, will do

Reply:
This conversation is covered by https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3106#discussion_r1799103083 now, will resolve it and leave the other open